benchmark/bench_util.c

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (C) the libgit2 contributors. All rights reserved.
+ *
+ * This file is part of libgit2, distributed under the GNU GPL v2 with
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+
+#include <fcntl.h>
+#include "common.h"
+#include "fileops.h"
+#include "buffer.h"
+
+static int tempdir_isvalid(const char *path)
+{
+	struct stat st;
+
+	if (p_stat(path, &st) != 0)
+		return 0;
+
+	if (!S_ISDIR(st.st_mode))
+		return 0;
+
+	return (access(path, W_OK) == 0);
+}
+
+int gitbench__create_tempdir(char **out)
+{
+	git_buf buf_env = GIT_BUF_INIT;
+	git_buf tempdir = GIT_BUF_INIT;
+	static const char *tempname = "libgit2_bench_XXXXXX";
+	int error = -1;
+
+	if ((git_buf_sets(&buf_env, getenv("BM_TEMP")) < 0) ||
```

Comment:
This helps on Windows with 260 character pathname problems.  One of the repos I was testing with has 240+ character paths and the default TMP/TEMP directory is somewhere down in %LOCALAPPDATA% (or whatever it is called) which expands to 70+ chars.  This would cause checkouts and merges to fail on that repo.

I was also using it for retargeting the benchmark to run on C:\t (SSD + shortnames) or D:\t (spinning rust without shortnames) and etc.

And it was nice to be able to do that without messing up my environment for other commands in the shell.

Also, it didn't see quite right using the CLAR_TEMP variable, but I could if you think that would be better.